nbt/src/main/java/net/kyori/adventure/nbt/ListBinaryTagImpl.java

```diff
@@ -64,34 +64,42 @@ public int size() {
   }
 
   @Override
-  public @NonNull ListBinaryTag set(final int index, final @NonNull BinaryTag newTag, final @Nullable Consumer<BinaryTag> removedConsumer) {
+  public @NonNull ListBinaryTag set(final int index, final @NonNull BinaryTag newTag, final @Nullable Consumer<? super BinaryTag> removed) {
     return this.edit(tags -> {
       final BinaryTag oldTag = tags.set(index, newTag);
-      if(removedConsumer != null) {
-        removedConsumer.accept(oldTag);
+      if(removed != null) {
+        removed.accept(oldTag);
       }
     }, newTag.type());
   }
 
   @Override
-  public @NonNull ListBinaryTag remove(final int index, final @Nullable Consumer<BinaryTag> removedConsumer) {
+  public @NonNull ListBinaryTag remove(final int index, final @Nullable Consumer<? super BinaryTag> removed) {
     return this.edit(tags -> {
-      final BinaryTag tag = tags.remove(index);
-      if(removedConsumer != null) {
-        removedConsumer.accept(tag);
+      final BinaryTag oldTag = tags.remove(index);
+      if(removed != null) {
+        removed.accept(oldTag);
       }
     }, null);
   }
 
   @Override
   public @NonNull ListBinaryTag add(final BinaryTag tag) {
+    noAddEnd(tag);
+    if(this.elementType != BinaryTagTypes.END) {
+      mustBeSameType(tag, this.elementType);
+    }
+    return this.edit(tags -> tags.add(tag), tag.type());
+  }
+
+  @Override
+  public @NonNull ListBinaryTag add(final Iterable<? extends BinaryTag> tagsToAdd) {
+    final BinaryTagType<?> type = ListBinaryTagImpl.mustBeSameType(tagsToAdd);
```

Comment:
add a check here that `tagsToAdd` is non-empty maybe, and if it is empty return `this` to save the copy?

Reply:
`Iterable` has no `isEmpty` or `size` method?

Reply:
you could either create an Iterator and check hasNext before going into `edit()`, or only do the optimization for inputs that are also a `Collection` i guess